[pt2][inductor] layout as part of cache key

nmacchioni opened the pull request:
Summary: include layout as part of the key, otherwise we don't have enough information to autotune offline (i.e. batch size for bmm is only in layout, not inputs)

Test Plan: sandcastle + CI

Differential Revision: D43525122



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @soumith @yanboliang @anijain2305 @desertfire

bertmaher reviewed (changes requested):
Code looks okay, but I don't understand the need for the change.  If I look at a random entry for `bmm`:
Old key:
```
"[('cuda', 'torch.float32', 2048, 160, 718, 114880, 1, 160, 0), ('cuda', 'torch.float32', 2048, 718, 16, 11488, 16, 1, 0)]"
```

New key:
```
"[('cuda', 'torch.float32', 2048, 160, 718, 114880, 1, 160, 0), ('cuda', 'torch.float32', 2048, 718, 16, 11488, 16, 1, 0)], FixedLayout('cuda', torch.float32, size=[2048, 160, 16], stride=[2560, 16, 1])"
```
It looks like the batch size *is* present in the old key: it's a `bmm([2048, 160, 718], [2048, 718, 16])`.  So I guess I don't understand what problem this diff is really solving

jansel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -609,6 +609,7 @@ def autotune(choice):
             choices,
             choices[0].name,
             repr([self.key_of(x) for x in input_nodes]),
+            repr(layout),
```
Add an:
```
assert isinsance(layout, ir.FixedLayout)
```
Want to make sure we aren't getting a FlexibleLayout here.

nmacchioni replied:
Just for clarification, is it the case that we shouldn't be getting ir.FlexibleLayout here or that we just shouldn't be including it in that case?

jansel replied:
If we are, it is a bug.

nmacchioni replied:
Ok I'll have to look into that, since we are getting flexible layouts here.

jansel replied:
You should just call `.freeze_layout()` to convert flexible->fixed.

jansel reviewed (commented):
Is this for output layouts?

nmacchioni commented:
> Code looks okay, but I don't understand the need for the change. If I look at a random entry for `bmm`: Old key:
> 
> ```
> "[('cuda', 'torch.float32', 2048, 160, 718, 114880, 1, 160, 0), ('cuda', 'torch.float32', 2048, 718, 16, 11488, 16, 1, 0)]"
> ```
> 
> New key:
> 
> ```
> "[('cuda', 'torch.float32', 2048, 160, 718, 114880, 1, 160, 0), ('cuda', 'torch.float32', 2048, 718, 16, 11488, 16, 1, 0)], FixedLayout('cuda', torch.float32, size=[2048, 160, 16], stride=[2560, 16, 1])"
> ```
> 
> It looks like the batch size _is_ present in the old key: it's a `bmm([2048, 160, 718], [2048, 718, 16])`. So I guess I don't understand what problem this diff is really solving

From my understanding the values in `repr(input)` are as follows: `[device, dtype, m, n, stride_m, stride_n, extra_size]`. I'm not sure if it is always the case that the stride is equal to the batch size.

nmacchioni commented:
> Is this for output layouts?

Yes, this was specifically for batched examples since I thought we might not have enough information in the inputs to actually reverse engineer the problem when autotuning offline.

nmacchioni committed:
[pt2][inductor] layout as part of cache key (#95359)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/95359

include layout as part of the key, otherwise we don't have enough information to autotune offline (i.e. batch size for bmm is only in layout, not inputs)

Test Plan: sandcastle + CI

Differential Revision: D43525122

fbshipit-source-id: 1f7cc4ce919fd16afa43550f0f789cda5df10148
